[SR] Add tooltips for disabled fields on managed SLM repository and policy

renshuki opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/173124#issuecomment-2352968634 by adding tooltips details when hovering the disabled SLM repository or policy fields.

### Screenshots

**SLM managed repository**
![Screenshot 2024-10-16 at 1 38 19](https://github.com/user-attachments/assets/3bd11ea5-f846-433f-8615-b51de184336b)

**SLM managed policy**
![Screenshot 2024-10-16 at 1 37 57](https://github.com/user-attachments/assets/d11757bd-bda5-4b4f-8c1e-e795e01b1fa2)

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#_add_your_labels)
- [x] This will appear in the **Release Notes** and follow the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)



<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

elasticmachine commented:
Pinging @elastic/kibana-management (Team:Kibana Management)

ElenaStoeva commented:
Thanks for working on this @renshuki! I'm adding the `ui-copy` label to have the copy reviewed. @elastic/kibana-design feel free to provide feedback on the design as well. Thanks!

florent-leborgne reviewed (approved):
The copy looks good to me, nice! 🌴

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/243168)
* Commit: b0a40843c4fc0658e3173cada1f8440b4631a434

### Failed CI Steps
* [Jest Tests #2](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9a07-4361-bfb3-3de44969ab80)
* [FTR Configs #30](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9a7b-43db-aae7-4138fe8250f6)
* [FTR Configs #50](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9ab5-4afe-819a-7eb09755c793)
* [FTR Configs #52](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9ab9-46e3-b204-14713d347833)
* [FTR Configs #53](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9abc-4825-b262-a7ec6541bbd3)
* [FTR Configs #57](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9ac6-413d-bedf-785c7e7f0da3)
* [FTR Configs #75](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9af2-45c2-bdfd-2b529fb06983)
* [FTR Configs #79](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9afc-4aac-9e98-09bfcccbecf3)
* [FTR Configs #88](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9b11-4975-9588-e308cf2e7a11)
* [FTR Configs #89](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9b13-454b-a54f-fbd4b5e6bfd6)
* [FTR Configs #108](https://buildkite.com/elastic/kibana-pull-request/builds/243168#01929607-9b41-4610-a5d2-afcf075aa578)
* [FTR Configs #108](https://buildkite.com/elastic/kibana-pull-request/builds/243168#0192962e-d91b-4112-a35c-1b9f22d9e0d8)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/243168","number":243168,"commit":"b0a40843c4fc0658e3173cada1f8440b4631a434"}],"number":243168}
buildkite-pr-comment-->

ElenaStoeva commented on x-pack/plugins/snapshot_restore/public/application/components/policy_form/steps/step_logistics.tsx:
```diff
@@ -54,6 +55,19 @@ export const PolicyStepLogistics: React.FunctionComponent<StepProps> = ({
+  const disableToolTip = (component: ReactElement): ReactElement => {
+    return policy?.isManagedPolicy ? (
+      <EuiToolTip
+        content="This field is disabled because you are editing a managed policy."
```
Can we add an i18n translation for this?

renshuki replied:
Added i18n translation for both (managed repository and managed policy)

ElenaStoeva commented on x-pack/plugins/snapshot_restore/public/application/components/repository_form/type_settings/azure_settings.tsx:
```diff
@@ -61,6 +62,19 @@ export const AzureSettings: React.FunctionComponent<Props> = ({
     });
   };
 
+  const disableToolTip = (component: ReactElement): ReactElement => {
```
Since this logic is used in multiple places, can we create a new React component under `snapshot_restore/public/application/components`  that takes `component` and `isManagedRepo` props and returns the corresponding component?

renshuki replied:
Thanks, I created a dedicated React component called DisableTooltip for the managed repository.

ElenaStoeva reviewed (changes requested):
Thanks for working on this @renshuki! I left a few comments regarding translations and code redundancy.

renshuki commented:
@ElenaStoeva thank you for the CR! I added i18n translations and created a React component for the managed repository.

I have two questions:

1. Should I create a dedicated React component for the managed policy as well, or is it acceptable to leave it as is since it's only used in one file?
2. Concerning translations, I've added `i18n.translate`. That being said, how should I handle translations for the different languages (`fr-FR`, `ja-JP`, `zh-CN`)?

Thank you 🙏

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/244485)
* Commit: c7a1fc0ffa7c03b897cc099afae1e4c9d3823f76

### Failed CI Steps
* [Jest Integration Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/244485#0192b1fa-ec32-4460-98ac-56b242d7a742)
* [FTR Configs #28](https://buildkite.com/elastic/kibana-pull-request/builds/244485#0192b1fa-ec64-4faa-8e3b-fd28de32ce04)
* [FTR Configs #51](https://buildkite.com/elastic/kibana-pull-request/builds/244485#0192b1fa-ec86-4308-8125-1863778901b2)
* [FTR Configs #51](https://buildkite.com/elastic/kibana-pull-request/builds/244485#0192b219-a549-4f9c-8f06-865e19db8b0d)

### History
* :broken_heart: [Build #243168](https://buildkite.com/elastic/kibana-pull-request/builds/243168) failed b0a40843c4fc0658e3173cada1f8440b4631a434

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/244485","number":244485,"commit":"c7a1fc0ffa7c03b897cc099afae1e4c9d3823f76"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/243168","number":243168,"commit":"b0a40843c4fc0658e3173cada1f8440b4631a434"}],"number":244485}
buildkite-pr-comment-->

ElenaStoeva commented on x-pack/plugins/snapshot_restore/public/application/components/policy_form/steps/step_logistics.tsx:
```diff
@@ -257,7 +273,7 @@ export const PolicyStepLogistics: React.FunctionComponent<StepProps> = ({
     }
 
-    return (
+    return disableToolTip(
```
Why don't we reuse the new `DisableToolTip` component here instead of this function?

renshuki replied:
It's because the tooltip message is different for the managed repository and the managed policy. But if we want to handle both cases (managed repository and managed policy) within the same component, I guess we can refactor the `DisableTooltip` component to something like:

**x-pack/plugins/snapshot_restore/public/application/components/disable_tooltip.tsx**

```js
export const DisableToolTip: React.FunctionComponent<Props> = ({
  isManagedRepository,
  isManagedPolicy,
  component,
}) => {
  let tooltipContent = '';

  if (isManagedRepository) {
    tooltipContent = i18n.translate('xpack.snapshotRestore.repositoryForm.disableToolTipContent', {
      defaultMessage: 'This field is disabled because you are editing a managed repository.',
    });
  } else if (isManagedPolicy) {
    tooltipContent = i18n.translate('xpack.snapshotRestore.policyForm.disableToolTipContent', {
      defaultMessage: 'This field is disabled because you are editing a managed policy.',
    });
  }

  return isManagedRepository || isManagedPolicy ? (
    <EuiToolTip content={tooltipContent} display="block">
      {component}
    </EuiToolTip>
  ) : (
    component
  );
};
```

Let me know what you think 😃

ElenaStoeva replied:
Ah I see... I would suggest that you pass the message as a prop to the component and rename the `isManagedRepository` prop to just `isManaged`. Something like this:
```
export const DisableToolTip: React.FunctionComponent<Props> = ({
  isManaged,
  tooltipMessage,
  component,
}) => {
  return isManaged ? (
    <EuiToolTip
      content={tooltipMessage}
      display="block"
    >
      {component}
    </EuiToolTip>
  ) : (
    component
  );
};
```

Also, could you please add documentation above this component to explain what we use it for?

renshuki replied:
and add a `const tooltipMessage` to define the tooltip message for each of the files, correct? (i.e. `azure_settings.tsx`, `gcs_settings.tsx`, `s3_settings.tsx` and `step_logistics.tsx`)

renshuki replied:
For example:
```
const tooltipMessage = i18n.translate('xpack.snapshotRestore.repositoryForm.disableToolTip', {
    defaultMessage: 'This field is disabled because you are editing a managed repository.',
  });
```

ElenaStoeva replied:
Yeah, I think it would be nice to add two constants (one for the repo message and one for the policy message) in `x-pack/plugins/snapshot_restore/public/application/constants/index.ts`. You could name them something more descriptive like `managed<Policy/Repository>FieldTooltipMessage`.

ElenaStoeva reviewed (commented):
Thanks for creating the tooltip wrapper component @renshuki - it looks good to me! I see it's been used in the repository form setting fields, but why don't we also use it for the managed policy logistics fields?

> Concerning translations, I've added i18n.translate. That being said, how should I handle translations for the different languages (fr-FR, ja-JP, zh-CN)?

No worries about adding the translations! As long as you add the translate message with a valid id, the translation would automatically be added after the PR is merged.

I'm removing the 8.16 label since the 8.16 FF date already passed, so let's target 8.17.

ElenaStoeva commented:
Also replaced `release_note:enhancement` label with `release_note:skip` as we usually only add to the Kibana release notes if the change is more substantial.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/245708)
* Commit: 5dd3b62cd02908a40be373c408cde95c18182585

### Failed CI Steps
* [Jest Integration Tests #3](https://buildkite.com/elastic/kibana-pull-request/builds/245708#0192bf6e-0459-4844-b939-93b94fc8dc53)
* [FTR Configs #33](https://buildkite.com/elastic/kibana-pull-request/builds/245708#0192bf6e-04a1-4544-8d60-d309b09b9e05)
* [FTR Configs #45](https://buildkite.com/elastic/kibana-pull-request/builds/245708#0192bf6e-04bc-425d-a8a9-854172f7a8f0)
* [FTR Configs #45](https://buildkite.com/elastic/kibana-pull-request/builds/245708#0192bf8c-a1d8-4a92-9ee7-b71c586ecabd)
* [FTR Configs #67](https://buildkite.com/elastic/kibana-pull-request/builds/245708#0192bf6e-04f1-42c7-87d8-287d6bde849e)

### History
* :broken_heart: [Build #244485](https://buildkite.com/elastic/kibana-pull-request/builds/244485) failed c7a1fc0ffa7c03b897cc099afae1e4c9d3823f76
* :broken_heart: [Build #243168](https://buildkite.com/elastic/kibana-pull-request/builds/243168) failed b0a40843c4fc0658e3173cada1f8440b4631a434

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245708","number":245708,"commit":"5dd3b62cd02908a40be373c408cde95c18182585"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/244485","number":244485,"commit":"c7a1fc0ffa7c03b897cc099afae1e4c9d3823f76"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/243168","number":243168,"commit":"b0a40843c4fc0658e3173cada1f8440b4631a434"}],"number":245708}
buildkite-pr-comment-->

ElenaStoeva commented on x-pack/plugins/snapshot_restore/public/application/constants/index.ts:
```diff
@@ -112,3 +114,18 @@ export const UIM_POLICY_CREATE = 'policy_create';
 export const UIM_RETENTION_EXECUTE = 'retention_execute';
+
+// DisableToolTip messages
+export const MANAGED_REPOSITORY_TOOLTIP_MESSAGE = i18n.translate(
```
I gave it a second thought and maybe these messages would be a better fit in `x-pack/plugins/snapshot_restore/public/application/components/disable_tooltip.tsx` since they would most likely be used only for the `DisableToolTip` component so it would make more sense to keep them together with the component. Apologies for the back-and-forth on this!

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/246008)
* Commit: ba592b9a20852e66cb49c91b418a9d6977445c2f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/0ff9a8a9d9ff2bdb99562eeca29152bd0a0c4385) | [after](https://github.com/elastic/kibana/commit/ba592b9a20852e66cb49c91b418a9d6977445c2f) | diff |
| --- | --- | --- | --- |
| `snapshotRestore` | 210 | 211 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/0ff9a8a9d9ff2bdb99562eeca29152bd0a0c4385) | [after](https://github.com/elastic/kibana/commit/ba592b9a20852e66cb49c91b418a9d6977445c2f) | diff |
| --- | --- | --- | --- |
| `snapshotRestore` | 261.5KB | 262.7KB | +1.1KB |



### History
* :broken_heart: [Build #245708](https://buildkite.com/elastic/kibana-pull-request/builds/245708) failed 5dd3b62cd02908a40be373c408cde95c18182585
* :broken_heart: [Build #244485](https://buildkite.com/elastic/kibana-pull-request/builds/244485) failed c7a1fc0ffa7c03b897cc099afae1e4c9d3823f76
* :broken_heart: [Build #243168](https://buildkite.com/elastic/kibana-pull-request/builds/243168) failed b0a40843c4fc0658e3173cada1f8440b4631a434

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/246008","number":246008,"commit":"ba592b9a20852e66cb49c91b418a9d6977445c2f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245708","number":245708,"commit":"5dd3b62cd02908a40be373c408cde95c18182585"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/244485","number":244485,"commit":"c7a1fc0ffa7c03b897cc099afae1e4c9d3823f76"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/243168","number":243168,"commit":"b0a40843c4fc0658e3173cada1f8440b4631a434"}],"number":246008}
buildkite-pr-comment-->

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/11522929432

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/197882">](https://github.com/elastic/kibana/pull/197882)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->

renshuki commented:
Thank you @ElenaStoeva for the code review back-and-forth and the great guidance you provided for this PR. It definitely helped me to learn more about the best practices and the Kibana code base. Thanks again! 🙇‍♂️
